iceoryx2-bb/log/src/logger/file.rs

```diff
@@ -0,0 +1,169 @@
+pub struct Logger {
+    state: Arc<Mutex<State>>,
+    trigger: Arc<Condvar>,
+    start_time: Instant,
+    _background_thread: Arc<Option<JoinHandle<()>>>,
+}
+
+impl Logger {
+    /// Creates a new file logger.
+    pub fn new(file_name: &str) -> Self {
+        let state = Arc::new(Mutex::new(State::default()));
+        let trigger = Arc::new(Condvar::new());
+        let mut file = OpenOptions::new()
+            .append(true)
+            .create(true)
+            .open(file_name)
+            .expect("Open log file for writing.");
+
+        let self_state = state.clone();
+        let self_trigger = trigger.clone();
+
+        let write_buffer_to_file = move || loop {
+            let lock = state.lock().expect("Acquire internal state mutex.");
+
+            let lock = trigger
+                .wait_while(lock, |state: &mut State| {
+                    state.buffer.is_empty() || !state.keep_running
+                })
+                .expect("Wait for internal state trigger.");
+            drop(lock);
+
+            loop {
+                // acquiring the lock only to get the next buffer to minimize contention as much
+                // as possible.
+                let mut lock = state.lock().expect("Acquire internal state mutex.");
+                let buffer = lock.buffer.pop_front();
+                drop(lock);
+
+                match buffer {
+                    Some(entry) => file
+                        .write_all(format!("{:?}\n", entry).as_bytes())
+                        .expect("Writing log message into log file."),
+                    None => break,
+                }
+            }
+            file.sync_all().expect("Sync log file with disc.");
+
+            let lock = state.lock().expect("Acquire internal state mutex.");
+            if !lock.keep_running {
+                break;
+            }
+        };
+
+        Self {
+            state: self_state,
+            trigger: self_trigger,
+            _background_thread: Arc::new(Some(std::thread::spawn(write_buffer_to_file))),
+            start_time: Instant::now(),
+        }
+    }
+}
+
+impl Drop for Logger {
+    fn drop(&mut self) {
+        let mut state = self.state.lock().expect("Acquire internal state mutex.");
+        state.keep_running = false;
+        self.trigger.notify_one();
+    }
+}
+
+impl crate::Logger for Logger {
+    fn log(
+        &self,
+        log_level: LogLevel,
+        origin: std::fmt::Arguments,
+        formatted_message: std::fmt::Arguments,
+    ) {
+        let mut state = self.state.lock().expect("Acquire internal state mutex.");
+        state.buffer.push_back(Entry {
+            log_level,
+            timestamp: SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("Acquire current system time."),
+            elapsed_time: self.start_time.elapsed(),
+            origin: origin.to_string(),
+            message: formatted_message.to_string(),
+        });
```

Comment:
If you want to keep the approach with the lock, please create the `Entry` before acquiring the lock to prevent large offsets in the timestamps to the actual log event.